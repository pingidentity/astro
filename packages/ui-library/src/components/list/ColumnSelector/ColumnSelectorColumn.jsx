import React, { Component } from "react";
import PropTypes from "prop-types";

const baseClassName = "column-selector__column";

export const ColumnTitle = ({
    subtitle,
    title
}) => (
    <div className={`${baseClassName}-title`}>
        {title}
        {subtitle && <div className={`${baseClassName}-subtitle`}>{subtitle}</div>}
    </div>
);

ColumnTitle.propTypes = {
    subtitle: PropTypes.node,
    title: PropTypes.node,
};

export default class ColumnSelectorColumn extends Component {
    static propTypes = {
        "data-id": PropTypes.string,
        subtitle: PropTypes.node,
        title: PropTypes.node,
    }

    static defaultProps = {
        "data-id": "column-selector-column",
    }

    render () {
        const {
            children,
            "data-id": dataId,
            title,
        } = this.props;

        return (
            <div className={baseClassName} data-id={dataId}>
                {title}
                <div className={`${baseClassName}-children`}>
                    {children}
                </div>
            </div>
        );
    }
}