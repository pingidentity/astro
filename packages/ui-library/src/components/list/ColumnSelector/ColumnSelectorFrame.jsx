import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import FormSearchBox from "../../forms/FormSearchBox";
import { isFunction, noop } from "underscore";

const baseClassName = "column-selector";

export default class ColumnSelector extends Component {
    static propTypes = {
        children: PropTypes.func.isRequired,
        className: PropTypes.string,
        "data-id": PropTypes.string,
        onSearch: PropTypes.func,
        searchPlaceHolder: PropTypes.string,
        query: PropTypes.string,
    };

    static defaultProps = {
        "data-id": "column-selector",
        onSearch: noop
    };

    state = {
        query: this.props.query || null
    }

    _handleSearch = query => {
        if (!this.props.query) {
            this.setState({ query });
        }

        this.props.onSearch(query);
    }

    render() {
        const {
            "data-id": dataId,
            children,
            className,
            searchPlaceHolder,
            query,
        } = this.props;

        return (
            <div className={classnames(baseClassName, className)} data-id={dataId}>
                <FormSearchBox
                    className={`${baseClassName}__search`}
                    data-id={`${dataId}-search`}
                    onValueChange={this._handleSearch}
                    placeholder={searchPlaceHolder}
                    queryString={query}
                />
                <div className={`${baseClassName}__columns`}>
                    {isFunction(children) ? children(this.state) : children}
                </div>
            </div>
        );
    }
}
