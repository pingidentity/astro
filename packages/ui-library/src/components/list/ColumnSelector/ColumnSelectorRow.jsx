import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import { isFunction, isString, noop } from "underscore";
import Button from "../../buttons/Button";
import FlexRow, { alignments } from "../../layout/FlexRow";
import Icon from "../../general/Icon";

export const buttonTypes = {
    ADD: "add",
    REMOVE: "remove"
};

const baseClassName = "column-selector__row";

export const RowButton = ({
    buttonType,
    dataId = `${baseClassName}-button`,
    expandable,
    onClick
}) => (
    <Button
        className={classnames(
            `${baseClassName}-button`,
            `${baseClassName}-button${expandable ? "--expandable" : ""}`
        )}
        data-id={`${dataId}-button`}
        iconName={buttonType === buttonTypes.ADD ? "plus" : "minus"}
        inline
        onClick={onClick}
    />
);

export const RowTitle = ({
    children
}) => (
    <div className={`${baseClassName}-title`}>
        {children}
    </div>
);

RowButton.propTypes = {
    buttonType: PropTypes.oneOf([
        buttonTypes.ADD,
        buttonTypes.REMOVE
    ]),
    "data-id": PropTypes.string,
    expandable: PropTypes.bool,
    onClick: PropTypes.func
};

RowButton.defaultProps = {
    "data-id": `column-selector-row-button`
};

export default class ColumnSelectorRow extends Component {
    static propTypes = {
        bottomPanel: PropTypes.node,
        buttonType: PropTypes.oneOf([
            buttonTypes.ADD,
            buttonTypes.REMOVE
        ]),
        className: PropTypes.string,
        customButton: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
        "data-id": PropTypes.string,
        disabled: PropTypes.bool,
        expandable: PropTypes.bool,
        id: PropTypes.string,
        onButtonClick: PropTypes.func,
        onToggle: PropTypes.func,
        open: PropTypes.oneOf([
            true,
            false,
            "initial"
        ]),
        subtitle: PropTypes.string,
        title: PropTypes.node
    }

    static defaultProps = {
        children: noop,
        className: "",
        "data-id": "column-selector-row",
        disabled: false,
        expandable: false,
        onButtonClick: noop,
        onToggle: noop
    }

    state = {
        open: (this.props.open === true || this.props.open === "initial") || false
    }

    _handleButtonClick = e => {
        // Stop propagation to avoid triggering click on parent components. May be null based
        // on custom button implementation
        if (e) {
            e.stopPropagation();
            e.persist();
        }

        this.props.onButtonClick({
            id: this.props.id,
            event: e
        });
    }

    _handleToggleOpen = e => {
        e.persist();

        const {
            id,
            onToggle
        } = this.props;

        const payload = {
            event: e,
            id,
            open: this.state.open
        };

        if (this.props.open === undefined || this.props.open === "initial") {
            this.setState(({ open }) => ({ open: !open }), () => onToggle(payload));
        } else {
            onToggle(payload);
        }
    }

    _renderButton = ({
        buttonType,
        customButton,
        disabled,
        draggable,
        expandable
    }) => {
        const button = this._renderCustomButton(customButton);
        if (button) {
            return button;
        }

        return (
            disabled || draggable
                ? <div />
                : <RowButton
                    buttonType={buttonType}
                    expandable={expandable}
                    onClick={this._handleButtonClick}
                />
        );
    }

    _renderCustomButton = button =>
        isFunction(button)
            ? button({
                handleOnButtonClick: this._handleButtonClick,
                ...this.props
            }, RowButton)
            : button

    render() {
        const {
            bottomPanel,
            children,
            className,
            "data-id": dataId,
            expandable,
            subtitle,
            title,
            titleIcon
        } = this.props;

        const renderedChildren = children(this.state);

        const expandableOpen = expandable && this.state.open;

        return (
            <div className={`${baseClassName}-container`} data-id={dataId}>
                <FlexRow
                    className={
                        classnames(
                            baseClassName,
                            className,
                            {
                                [`${baseClassName}--open`]: expandableOpen,
                                [`${baseClassName}--closed`]: expandable && !this.state.open,
                                [`${baseClassName}--disabled`]: this.props.disabled
                            }
                        )}
                    inline
                    alignment={alignments.STRETCH}
                >
                    {expandable
                        ? <div
                            className={`${baseClassName}-toggle-container`}
                            data-id="row-toggle"
                            onClick={this._handleToggleOpen}
                        >
                            <Icon
                                className={`${baseClassName}-toggle`}
                                data-id={`${dataId}-toggle-icon`}
                                iconName={this.state.open ? "close-arrow" : "dropdown-arrow"}
                                type="inline"
                            />
                        </div>
                        : <div className={`${baseClassName}-title-icon`}>
                            {titleIcon &&
                                <Icon
                                    iconName={titleIcon}
                                    type="inline"
                                />
                            }
                        </div>
                    }
                    <div
                        className={
                            classnames(
                                `${baseClassName}-content`,
                                {
                                    [`${baseClassName}-content--open`]: expandableOpen
                                }
                            )
                        }
                    >
                        <div className={`${baseClassName}-titles`}>
                            {isString(title) ? <RowTitle>{title}</RowTitle> : title}
                            {subtitle &&
                            <div className={`${baseClassName}-subtitle`}>
                                {subtitle}
                            </div>
                            }
                            {bottomPanel &&
                            <div className={`${baseClassName}-titles-panel`}>
                                {bottomPanel}
                            </div>
                            }
                        </div>
                        {this._renderButton(this.props)}
                    </div>
                </FlexRow>
                {renderedChildren && <div className={`${baseClassName}-options`}>
                    {renderedChildren}
                </div>}
            </div>
        );
    }
}