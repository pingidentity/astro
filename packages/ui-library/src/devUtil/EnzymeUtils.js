import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
