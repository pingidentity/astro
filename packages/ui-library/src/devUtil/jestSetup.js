
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
