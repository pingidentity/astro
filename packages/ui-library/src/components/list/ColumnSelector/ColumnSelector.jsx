import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import { noop, partition } from "underscore";
import ColumnSelectorColumn from "./ColumnSelectorColumn";
import ColumnSelectorFrame from "./ColumnSelectorFrame";
import ColumnSelectorRow, { buttonTypes } from "./ColumnSelectorRow";
import DragDrop from "../../rows/DragDrop";

const baseClassName = "column-selector";

/**
 * @class Column Selector
 * @desc A component that displays two columns; options can be moved between them.
 *
 * @param {string} [data-id="column-selector"]
 *     To define the base "data-id" value for top-level HTML container.
 * @param {string} [className]
 *     CSS classes to set on the top-level HTML container.
 * @param {function} [customSort]
 *     A custom sort function that the column selector will run to sort columns.
 *     The function will be passed to an Array.sort function.
 *
 * @param {boolean} [draggable=false]
 *     Determines whether rows can be dragged to add and rearrange. Not fully
 *     implemented.
 *
 * @callback onAdd
 *     Callback for when the Add button is clicked for an option.
 * @param {Object} payload
 *     An object that holds all of the callback information for onAdd.
 * @param {string} payload.id
 *     ID of the object being added.
 * @param {Object[]} payload.options
 *     The options with the added option removed.
 * @param {Object[]} payload.selectedOptions
 *     The selectedOptions with the added option added.
 * @param {Object[]} payload.event
 *     The JS even that triggered the add.
 *
 * @callback onRemove
 *     Callback for when the Remove button is clicked for a selected option.
 * @param {Object} payload
 *     An object that holds all of the callback information for onRemove.
 * @param {string} payload.id
 *     ID of the option being removed.
 * @param {Object[]} payload.options
 *     The options with the remove option added.
 * @param {Object[]} payload.selectedOptions
 *     The selectedOptions with the removed option removed.
 * @param {Object[]} payload.event
 *     The JS even that triggered the add.
 *
 * @callback onSearch
 *     Callback for when an expandable option is opened or closed.
 * @param {string} query
 *     The search query currently entered in the search box.
 *
 * @callback onToggleOption
 *     Callback for when an expandable option is opened or closed.
 * @param {Object} payload
 *     An object that holds all of the callback information for onRemove.
 * @param {string} payload.id
 *     ID of the option being toggled.
 * @param {Object[]} payload.event
 *     The JS even that triggered the add.
 * @param {boolean} openParentOnMove=true
 *     If true, open rows in options or selectedOptions with the same ID will
 *     cause the option with the same ID in the other section to also be open.
 *
 * @param {Object[]} options
 *     The options shown on the left-hand side of the component; can be added.
 * @param {string} options[].id
 *     The id of the option.
 * @param {Object[]} options[].children
 *     The child options of an option; makes option expandable. Params are same as parent,
 *     except children cannot have children.
 * @param {Object} options[].customButton
 *     A JSX node that replaces the add button.
 * @param {boolean} options[].open
 *     Determines whether object is open or closed; only valid if object has children.
 *     Option will manage its own open state if not passed in.
 * @param {string} subtitle
 *     String to be displayed below title. Defaults to parent title for children.
 * @param {Object} title
 *     Title for the option. Accepts JSX or a string.
 * @param {string} titleIcon
 *     Icon to be displayed next to the title. Does not render for expandable options.
 *
 * @param {Object} optionsTitle
 *     What is displayed above options. Takes JSX or a string.
 * @param {string} searchPlaceholder
 *     Placeholder for the search bar.
 * @param {Object[]} selectedOptions
 *     Options to be displayed on right side of selected; can be removed. Properties are same as
 *     options prop, listed above.
 * @param {Object} selectedTitle
 *     What is displayed above selected options. Takes JSX or a string.
 *
 * @param {Search~onResultClick}
 *     Callback to be triggered when a result is clicked; passes back result properties from possibleResults
 *
 */
export default class ColumnSelector extends Component {
    static propTypes = {
        className: PropTypes.string,
        customSort: PropTypes.func,
        "data-id": PropTypes.string,
        draggable: PropTypes.bool,
        onAdd: PropTypes.func,
        onRemove: PropTypes.func,
        onSearch: PropTypes.func,
        onToggleOption: PropTypes.func,
        openParentOnMove: PropTypes.bool,
        options: PropTypes.arrayOf(PropTypes.shape({
            id: PropTypes.string.isRequired,
            children: PropTypes.arrayOf(
                PropTypes.shape({
                    id: PropTypes.string.isRequired,
                    customButton: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
                    subtitle: PropTypes.string,
                    title: PropTypes.node,
                    titleIcon: PropTypes.string
                })
            ),
            customButton: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
            open: PropTypes.bool,
            subtitle: PropTypes.string,
            title: PropTypes.node,
            titleIcon: PropTypes.string
        })).isRequired,
        optionsTitle: PropTypes.node,
        searchPlaceHolder: PropTypes.string,
        selectedOptions: PropTypes.arrayOf(PropTypes.shape({
            all: PropTypes.bool,
            id: PropTypes.string.isRequired,
            label: PropTypes.node,
            children: PropTypes.arrayOf(
                PropTypes.shape({
                    id: PropTypes.string.isRequired,
                    customButton: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
                    subtitle: PropTypes.string,
                    title: PropTypes.node,
                    titleIcon: PropTypes.string
                })
            ),
            customButton: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
            open: PropTypes.bool,
            subtitle: PropTypes.string,
            title: PropTypes.node,
            titleIcon: PropTypes.string
        })),
        selectedTitle: PropTypes.node
    };

    static defaultProps = {
        "data-id": "column-selector",
        draggable: false,
        onAdd: noop,
        onFilter: noop,
        onRemove: noop,
        onSearch: noop,
        onToggleOption: noop,
        openParentOnMove: true,
        selectedOptions: []
    };

    state = {
        lastMoved: undefined
    }

    addDragDrop = (row, isSelected) =>
        this.props.draggable
            ? this.renderDragDrop(row, isSelected)
            : row

    renderDragDrop = (row, isSelected) => {
        const {
            id
        } = row.props;

        // Skipping due to complexity of testing drag and drop
        /* istanbul ignore next  */
        const onDrop = (...args) => {
            return isSelected
                ? this.props.onAdd({
                    id: args[1]
                })
                : this.props.onRemove({
                    id: args[1]
                });
        };

        return (
            <DragDrop
                column={isSelected ? 1 : 0}
                id={id}
                index={id}
                key={id}
                onDrop={onDrop}
            >
                {row}
            </DragDrop>
        );
    }

    // Creates a placeholder drag target
    renderDropPlaceholder = isSelected => {
        // Skipping due to complexity of testing drag and drop
        /* istanbul ignore next  */
        const onDrop = (...args) =>
            isSelected
                ? this.props.onAdd({
                    id: args[1]
                })
                : this.props.onRemove({
                    id: args[1]
                });

        return (
            <DragDrop
                className={`${baseClassName}__drop-placeholder`}
                column={isSelected ? 1 : 0}
                id={isSelected ? "selected-placeholder" : "base-placeholder"}
                index="placeholder"
                key="drop-placeholder"
                onDrop={onDrop}
                removeDraggableAttribute
            />
        );
    }

    handleAdd = ({ id, ...payload }) => {
        const [added, options] = partition(this.props.options, (({ id: optId }) => id === optId));
        if (this.props.openParentOnMove) {
            this.setState({
                lastMoved: id
            }, () => this.props.onAdd({
                id,
                options,
                selectedOptions: [...this.props.selectedOptions, ...added],
                ...payload
            }));
        } else {
            this.props.onAdd({
                id,
                options,
                selectedOptions: [...this.props.selectedOptions, ...added],
                ...payload
            });
        }
    }
    handleRemove = ({ id, ...payload }) => {
        const [removed, selected] = partition(this.props.selectedOptions, (({ id: optId }) => id === optId));
        if (this.props.openParentOnMove) {
            this.setState({
                lastMoved: id
            }, () => this.props.onRemove({
                id,
                options: [...this.props.options, ...removed],
                selectedOptions: selected,
                ...payload
            }));
        } else {
            this.props.onRemove({
                id,
                options: [...this.props.options, ...removed],
                selectedOptions: selected,
                ...payload
            });
        }
    }
    handleToggleOption = (id, open) => this.props.onToggleOption({ id, open });

    // Curried so that it can be given an arbitrary button type
    renderChild = (buttonType, handleButtonClick, isSelected) => ({
        id,
        ...child
    }) => (
        this.addDragDrop(
            <ColumnSelectorRow
                buttonType={buttonType}
                id={id}
                key={id}
                onButtonClick={handleButtonClick}
                {...child}
            />
        , isSelected)
    );

    // Renders options; curried so that it can pass an icon to all of its children, decide whether
    // open should remain undefined or converted to false, and to be passed to a map function
    renderOption = (buttonType, handleButtonClick, manageOpen, isSelected) => ({
        children = [],
        id,
        open = manageOpen ? undefined : false,
        title,
        ...option
    }) => {
        const render = this.renderChild(buttonType, handleButtonClick, isSelected);
        const initialOpen =
            this.props.openParentOnMove
                ? (children.some(({ id: childId }) => childId === this.state.lastMoved) && "initial") || undefined
                : undefined;

        return this.addDragDrop(
            <ColumnSelectorRow
                buttonType={buttonType}
                data-id={`${this.props["data-id"]}-${id}`}
                expandable={children.length > 0}
                id={id}
                key={id}
                onButtonClick={handleButtonClick}
                onToggle={this.handleToggleOption}
                open={manageOpen ? initialOpen : open}
                title={title}
                {...option}
            >
                {({ open: stateOpen }) => {
                    if (stateOpen === true) {
                        return children.map(child => render({
                            subtitle: title,
                            ...child
                        }));
                    } else {
                        return null;
                    }
                }}
            </ColumnSelectorRow>
        , isSelected);
    };

    shouldManageOpen = options => !options.some(({ open }) => open !== undefined);

    shouldManageOpenOptions = this.shouldManageOpen(this.props.options)
    shouldManageOpenSelected = this.shouldManageOpen(this.props.selectedOptions)

    renderBaseOption =
        this.renderOption(
            buttonTypes.ADD,
            this.handleAdd,
            this.shouldManageOpenOptions,
            false
        )

    renderSelectedOption =
        this.renderOption(
            buttonTypes.REMOVE,
            this.handleRemove,
            this.shouldManageOpenSelected,
            true
        )

    sortOptions = options => this.props.customSort ? options.sort(this.props.customSort) : options;

    render() {
        const {
            "data-id": dataId,
            className,
            onSearch,
            options,
            optionsTitle,
            searchPlaceHolder,
            selectedTitle,
            selectedOptions,
        } = this.props;

        const sortedOptions = this.sortOptions(options);
        const sortedSelected = this.sortOptions(selectedOptions);

        return (
            <ColumnSelectorFrame
                dataId={dataId}
                className={classnames(baseClassName, className)}
                onSearch={onSearch}
                searchPlaceHolder={searchPlaceHolder}
            >
                {() => [
                    <ColumnSelectorColumn
                        data-id={`${dataId}-options`}
                        key="options"
                        title={optionsTitle}
                    >
                        {[
                            ...sortedOptions.map(this.renderBaseOption),
                            this.renderDropPlaceholder(false)
                        ]}
                    </ColumnSelectorColumn>,
                    <ColumnSelectorColumn
                        data-id={`${dataId}-selected`}
                        key="selected"
                        title={selectedTitle}
                    >
                        {[
                            ...sortedSelected.map(this.renderSelectedOption),
                            this.renderDropPlaceholder(true)
                        ]}
                    </ColumnSelectorColumn>
                ]}
            </ColumnSelectorFrame>
        );
    }
}
