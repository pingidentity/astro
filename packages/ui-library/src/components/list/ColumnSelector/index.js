import ColumnSelector from "./ColumnSelector";
import ColumnSelectorColumn, { ColumnTitle } from "./ColumnSelectorColumn";
import ColumnSelectorRow, { buttonTypes, RowButton, RowTitle } from "./ColumnSelectorRow";

ColumnSelector.buttonTypes = buttonTypes;
ColumnSelector.ColumnSelectorColumn = ColumnSelectorColumn;
ColumnSelector.ColumnSelectorRow = ColumnSelectorRow;
ColumnSelector.ColumnTitle = ColumnTitle;
ColumnSelector.RowButton = RowButton;
ColumnSelector.RowTitle = RowTitle;

export default ColumnSelector;