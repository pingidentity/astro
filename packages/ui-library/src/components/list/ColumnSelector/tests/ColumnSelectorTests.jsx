import React from "react";
import { mount, shallow } from "enzyme";
import { omit } from "underscore";
import { mountSnapshotDataIds } from "../../../../devUtil/EnzymeUtils";
import ColumnSelector from "../ColumnSelector";
import ColumnSelectorFrame from "../ColumnSelectorFrame";
import ColumnSelectorRow, { RowButton } from "../ColumnSelectorRow";

window.__DEV__ = true;

jest.mock("../../../rows/DragDrop.jsx", () => () => <div className="drag-drop-mock" />);

describe("ColumnSelector", () => {
    const baseOptions = [
        {
            id: "prod",
            subtitle: "US, Europe and Israel",
            title: "Production",
            //Can add "open" property here; the component will not manage state for whether options
            //are collapsed or expanded if any options have this property defined.
            children: [
                {
                    id: "prod-us-employees",
                    title: "US Employees",
                    titleIcon: "globe"
                },
                {
                    id: "prod-europe-employees",
                    title: "Europe Employees",
                    titleIcon: "globe"
                },
                {
                    id: "prod-israel-employees",
                    title: "Israel Employees",
                    titleIcon: "globe"
                },
            ],
        },
        {
            id: "sandbox",
            subtitle: "US, Europe and Israel",
            title: "Sandbox",
            children: [
                {
                    id: "sandbox-us-employees",
                    title: "US Employees",
                    titleIcon: "globe"
                },
                {
                    id: "sandbox-europe-employees",
                    title: "Europe Employees",
                    titleIcon: "globe"
                },
                {
                    id: "sandbox-israel-employees",
                    title: "Israel Employees",
                    titleIcon: "globe"
                },
            ]
        },
        {
            id: "test",
            subtitle: "US, Europe and Israel",
            title: "Test",
            children: [
                {
                    id: "test-us-employees",
                    title: "US Employees",
                    titleIcon: "globe"
                },
                {
                    id: "test-europe-employees",
                    title: "Europe Employees",
                    titleIcon: "globe"
                },
                {
                    id: "test-israel-employees",
                    title: "Israel Employees",
                    titleIcon: "globe"
                },
            ],
            customButton: ({ handleOnButtonClick }) => {
                //handleOnButtonClick();

                return (<RowButton
                    onClick={handleOnButtonClick}
                />);
            }
        }
    ];

    function getComponent({
        "dataid": dataId = "test-selector",
        options = baseOptions,
        ...props
    } = {}) {
        return shallow(
            <ColumnSelector
                data-id={dataId}
                options={options}
                selectedOptions={baseOptions}
                {...props}
            />
        );
    }

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <ColumnSelector
                options={baseOptions}
                selectedOptions={baseOptions}
            />
        );
    });

    it("renders the first column", () => {
        const component = getComponent();

        // Make sure that render prop fires
        const [firstColumn] =
            component
                .find(ColumnSelectorFrame)
                .props()
                .children;
        const renderedColumn = shallow(firstColumn);

        expect(renderedColumn.exists()).toEqual(true);
    });

    it("renders the first row", () => {

    });

    it("renders row with no children", () => {
        const component = getComponent({
            options: baseOptions.map(opt => omit(opt, "children"))
        });

        // Make sure that render prop fires
        const [firstColumn] =
            component
                .find(ColumnSelectorFrame)
                .props()
                .children;

        const [firstRow] =
            shallow(firstColumn)
                .find(ColumnSelectorRow);

        const renderedRow = mount(firstRow);

        expect(renderedRow.props().children({})).toEqual(null);
    });

    it("renders closed row with children", () => {
        const component = getComponent({
            options: [
                {
                    ...baseOptions[0],
                    open: false
                }
            ]
        });

        // Make sure that render prop fires
        const [firstColumn] =
            component
                .find(ColumnSelectorFrame)
                .props()
                .children;

        const [firstRow] =
            shallow(firstColumn)
                .find(ColumnSelectorRow);

        const renderedRow = mount(firstRow);

        expect(renderedRow.props().children({})).toEqual(null);
    });

    it("renders open row with children", () => {
        const component = getComponent({
            options: [
                {
                    ...baseOptions[0],
                    open: true
                }
            ]
        });

        // Make sure that render prop fires
        const [firstColumn] =
            component
                .find(ColumnSelectorFrame)
                .props()
                .children;

        const [firstRow] =
            shallow(firstColumn)
                .find(ColumnSelectorRow);

        const childRows = shallow(firstRow).find(ColumnSelectorRow);

        expect(childRows.length).toEqual(3);
    });

    it("renders drag drop rows if draggable prop is true", () => {
        const component = mount(
            <ColumnSelector
                draggable
                options={baseOptions}
            />
        );

        const draggable = component.find("div.drag-drop-mock");
        expect(draggable.exists()).toEqual(true);
    });

    it("does not render drag drop rows if draggable prop is false", () => {
        const openOptions = baseOptions.map(opt => ({ ...opt, open: true }));
        const component = mount(
            <ColumnSelector
                draggable={false}
                options={openOptions}
                selectedOptions={openOptions}
            />
        );

        const draggable = component.find("div.drag-drop-mock");
        expect(draggable.exists()).toEqual(true);
    });

    it("handleAdd calls onAdd with options and selectedOptions added to payload", () => {
        const onAdd = jest.fn();
        const component = getComponent({
            onAdd,
            options: [],
            selectedOptions: []
        });
        component.instance().handleAdd({});

        expect(onAdd).toHaveBeenCalledWith({
            options: [],
            selectedOptions: []
        });
    });

    it("handleAdd correctly sets lastMoved if openSameId prop is true", () => {
        const component = getComponent({
            options: baseOptions,
            selectedOptions: baseOptions
        });
        component.instance().handleAdd({ id: "sandbox-israel-employees" });

        expect(component.state().lastMoved).toEqual("sandbox-israel-employees");
    });

    it("handleAdd does not set lastMoved if openSameId prop is false", () => {
        const component = getComponent({
            openParentOnMove: false,
            options: baseOptions,
            selectedOptions: baseOptions
        });
        component.instance().handleAdd({ id: "sandbox-israel-employees" });

        expect(component.state().lastMoved).not.toEqual("sandbox-israel-employees");
    });

    it("handleRemove correctly sets lastMoved if openSameId prop is true", () => {
        const component = getComponent({
            options: baseOptions,
            selectedOptions: baseOptions
        });
        component.instance().handleRemove({ id: "sandbox-israel-employees" });

        expect(component.state().lastMoved).toEqual("sandbox-israel-employees");
    });

    it("handleRemove does not set lastMoved if openSameId prop is false", () => {
        const component = getComponent({
            openParentOnMove: false,
            options: baseOptions,
            selectedOptions: baseOptions
        });
        component.instance().handleRemove({ id: "sandbox-israel-employees" });

        expect(component.state().lastMoved).not.toEqual("sandbox-israel-employees");
    });

    it("handleRemove calls onRemove with options and selectedOptions added to payload", () => {
        const onRemove = jest.fn();
        const component = getComponent({
            onRemove,
            options: [],
            selectedOptions: []
        });
        component.instance().handleRemove({});

        expect(onRemove).toHaveBeenCalledWith({
            options: [],
            selectedOptions: []
        });
    });

    it("handleToggleOptions calls onToggleOption with options and selectedOptions added to payload", () => {
        const onToggleOption = jest.fn();
        const component = getComponent({
            onToggleOption,
        });
        component.instance().handleToggleOption("test id", true);

        expect(onToggleOption).toHaveBeenCalledWith({
            id: "test id",
            open: true
        });
    });

    it("sorts rows using customSort function", () => {
        const sort = ({ id: idA }, { id: idB }) => idA > idB ? 1 : -1;
        const options = [
            {
                id: "second",
                title: "second"
            },
            {
                id: "third",
                title: "third"
            },
            {
                id: "first",
                title: "first"
            }
        ];

        const component = mount(
            <ColumnSelector
                customSort={sort}
                draggable={false}
                options={options}
            />
        );

        const [firstRow] = component.find(ColumnSelectorRow);

        expect(mount(firstRow).props().id).toEqual("first");
    });
});
