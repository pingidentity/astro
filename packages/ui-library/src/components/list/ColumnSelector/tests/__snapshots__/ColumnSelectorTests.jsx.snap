// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnSelector data-id's don't change 1`] = `
Array [
  "column-selector",
  "column-selector-search",
  "searchBox",
  "text-field_container",
  "searchBox-input",
  "column-selector-options",
  "column-selector-prod",
  "flex-row",
  "row-toggle",
  "column-selector-prod-toggle-icon",
  "column-selector__row-button-button",
  "column-selector-sandbox",
  "flex-row",
  "row-toggle",
  "column-selector-sandbox-toggle-icon",
  "column-selector__row-button-button",
  "column-selector-test",
  "flex-row",
  "row-toggle",
  "column-selector-test-toggle-icon",
  "column-selector__row-button-button",
  "column-selector-selected",
  "column-selector-prod",
  "flex-row",
  "row-toggle",
  "column-selector-prod-toggle-icon",
  "column-selector__row-button-button",
  "column-selector-sandbox",
  "flex-row",
  "row-toggle",
  "column-selector-sandbox-toggle-icon",
  "column-selector__row-button-button",
  "column-selector-test",
  "flex-row",
  "row-toggle",
  "column-selector-test-toggle-icon",
  "column-selector__row-button-button",
]
`;
